internal/internal_update.go

```diff
@@ -207,37 +217,47 @@ func defaultUpdateHandler(
 	serializedArgs *commonpb.Payloads,
 	header *commonpb.Header,
 	callbacks UpdateCallbacks,
-	spawn func(Context, string, func(Context)) Context,
+	scheduler UpdateScheduler,
 ) {
 	env := getWorkflowEnvironment(rootCtx)
 	ctx, err := workflowContextWithHeaderPropagated(rootCtx, header, env.GetContextPropagators())
 	if err != nil {
 		callbacks.Reject(err)
 		return
 	}
-	eo := getWorkflowEnvOptions(ctx)
-	handler, ok := eo.updateHandlers[name]
-	if !ok {
-		keys := make([]string, 0, len(eo.updateHandlers))
-		for k := range eo.updateHandlers {
-			keys = append(keys, k)
+	scheduler.Spawn(ctx, name, func(ctx Context) {
+		eo := getWorkflowEnvOptions(ctx)
+
+		// If we suspect that handler registration has not occurred (e.g.
+		// because this update is part of the first workflow task and is being
+		// delivered before the workflow function itself has run and had a
+		// chance to register update handlers) then we yield control back to the
+		// scheduler to allow handler registration to occur. The sceduler will
+		// resume this coroutine after others have run to a blocking point.
```

Comment:
Hrmm. I am racking my brain trying to think of a downside to this. We will have to discuss what update-with-start looks like I think (just whether the update event coming before the workflow start event is even an issue). 

I can't think of any downsides to this off the top of my head. Even if there are multiple updates before workflow start I think a yield from each is safe. And I don't think yielding in every case of no handlers no matter how far into the workflow we are is harmful either (just pushing this to the end of the loop). Is there any change this yield could not get completed before next WFT? I know there's some "run until all yielded" code/logic before commands are sent as a WFT completion, so I want to confirm that this "yield" is not considered waiting on Temporal like the other yields.

I would like to request a update-before-worker-start integration test in the features repo (though all other languages offer pre-start handler registration unlike Go).

Reply:
The downside for me is that this aspect of update delivery (admittedly an edge-case) becomes dependent on exactly how workflow execution happens. In particular on the fact that during startup the root coro yields exactly once before running the user-supplied workflow task.

As for the update-before-start integration test ... that's actually all of them! Prior to this PR, I always had to time.Sleep(1*time.Second) between launching the workflow and submitting the first update. Without that sleep, the first update sees an error returned "no registered handler for 'foo'" With this change, that sleep can be removed.

Reply:
> The downside for me is that this aspect of update delivery (admittedly an edge-case) becomes dependent on the precise mechanism of how workflow execution happens and it does so in a somewhat invisible way.

This is the accepted cost of Go SDK's inability to predefine anything about a workflow before starting it. No worries (much cleaner in other langs)

> is not considered to have blocked [...] within the current call to ExecuteUntilAllBlocked 

Perfect, thanks